src/safeds_runner/server/_memoization_utils.py

```diff
@@ -0,0 +1,464 @@
+        _set_new_explicit_memory(self.value, self.memory)
+
+
+@dataclass(frozen=True)
+class ExplicitIdentityWrapperLazy:
+    """
+    Wrapper containing a value that lives in a shared memory location, and supports a deterministic hash.
+
+    This wrapper allows to skip deserializing the contained value, if only a comparison is required, as the hash is deterministic and also sent.
+    If the comparison using the explicit identity fails, the object is unpickled as a fallback solution and compared using the __eq__ function.
+    """
+
+    value: Any
+    memory: SharedMemory
+    id: uuid.UUID
+    hash: int
+
+    @classmethod
+    def shared(cls, value: Any) -> ExplicitIdentityWrapperLazy:
+        """
+        Create a new wrapper around the provided value.
+
+        The provided value will be pickled and stored in shared memory, for storage in the memoization cache.
+        An explicit identity should already be assigned, and the object should be deterministically hashable.
+
+        Parameters
+        ----------
+        value
+            Object to create a shared memory based wrapper for.
+
+        Returns
+        -------
+        result
+            A new wrapper object containing the provided value.
+        """
+        _shared_memory_serialize_and_assign(value)
+        return cls.existing(value)
+
+    @classmethod
+    def existing(cls, value: Any) -> ExplicitIdentityWrapperLazy:
+        """
+        Create a wrapper around the provided value, by using the existing assigned shared memory location.
+
+        Parameters
+        ----------
+        value
+            Object to create a shared memory based wrapper for.
+
+        Returns
+        -------
+        result
+            A new wrapper object containing the provided value.
+        """
+        return cls(value, value.__ex_id_mem__, value.__ex_id__, value.__ex_hash__)
+
+    def __eq__(self, other: object) -> bool:
+        if (
+            isinstance(other, ExplicitIdentityWrapperLazy)
+            and self.id == other.id
+            or isinstance(other, ExplicitIdentityWrapper)
+            and self.id == other.value.__ex_id__
+            or _has_explicit_identity(other)
+            and self.id == other.__ex_id__  # type: ignore[attr-defined]
+        ):
+            return True
+        self._unpackvalue()
+        if isinstance(other, ExplicitIdentityWrapperLazy):
+            other._unpackvalue()
+            return self.value == other.value
+        elif isinstance(other, ExplicitIdentityWrapper):
+            return self.value == other.value
+        return self.value == other
+
+    def __hash__(self) -> int:
+        return self.hash
+
+    def _unpackvalue(self) -> None:
+        """Unpack the value contained in this wrapper, if not currently present."""
+        if self.value is None:
+            object.__setattr__(self, "value", pickle.loads(self.memory.buf))
+            _set_new_explicit_memory(self.value, self.memory)
```

Comment:
How about 

1. Renaming the field `value` to `_value`
2. Adding a `@property` `value` that unpacks the value and stores in in `self_value` if `self_value` is `None` and then returns `self._value`.

This way callers don't need to remember unpacking the value first, but can just access the value directly.